pallets/nodle-staking/src/migrations.rs

```diff
@@ -0,0 +1,297 @@
+#[allow(dead_code)]
+pub fn poa_validators_migration<T: Config>() -> Weight {
+    // Buffer to account for misc checks
+    let mut weight: Weight = 1_000;
+
+    log::info!("🕊️ Starting Poa validators migration...");
+
+    struct POAValidatorsStorageValuePrefix;
+    impl StorageInstance for POAValidatorsStorageValuePrefix {
+        const STORAGE_PREFIX: &'static str = "Validators";
+        fn pallet_prefix() -> &'static str {
+            "Poa"
+        }
+    }
+
+    struct StakingValidatorStateStorageMapPrefix;
+    impl StorageInstance for StakingValidatorStateStorageMapPrefix {
+        const STORAGE_PREFIX: &'static str = "ValidatorState";
+        fn pallet_prefix() -> &'static str {
+            "NodleStaking"
+        }
+    }
+
+    struct StakingInvulnerablesStorageValuePrefix;
+    impl StorageInstance for StakingInvulnerablesStorageValuePrefix {
+        const STORAGE_PREFIX: &'static str = "Invulnerables";
+        fn pallet_prefix() -> &'static str {
+            "NodleStaking"
+        }
+    }
+
+    #[allow(type_alias_bounds)]
+    type POAValidators<T: Config> =
+        StorageValue<POAValidatorsStorageValuePrefix, Vec<T::AccountId>, OptionQuery>;
+
+    #[allow(type_alias_bounds)]
+    type StakingInvulnerables<T: Config> =
+        StorageValue<StakingInvulnerablesStorageValuePrefix, Vec<T::AccountId>, OptionQuery>;
+
+    #[allow(type_alias_bounds)]
+    type StakingValidatorState<T: Config> = StorageMap<
+        StakingValidatorStateStorageMapPrefix,
+        Twox64Concat,
+        T::AccountId,
+        Validator<T::AccountId, BalanceOf<T>>,
+        OptionQuery,
+    >;
+
+    log::info!(
+        "<POAValidators<T>>::exits()? {:?}",
+        <POAValidators<T>>::exists()
+    );
+    log::info!(
+        "<StakingInvulnerables<T>>::exits()? {:?}",
+        <StakingInvulnerables<T>>::exists()
+    );
+
+    if !<POAValidators<T>>::exists() {
+        log::info!("🕊️ POA Validators doesn't exist Nothing to migrate!!!");
+    } else {
+        let poa_validators = <POAValidators<T>>::get().unwrap();
+        weight = weight.saturating_add(RocksDbWeight::get().reads_writes(1, 1));
+
+        let pre_invulnerable_validators = <StakingInvulnerables<T>>::get().unwrap();
+        weight = weight.saturating_add(RocksDbWeight::get().reads_writes(1, 1));
+
+        let mut post_invulnerable_validators = pre_invulnerable_validators
+            .clone()
+            .iter()
+            .chain(poa_validators.iter())
+            .map(|x| x.clone())
+            .collect::<Vec<T::AccountId>>();
+
+        post_invulnerable_validators.sort();
+        post_invulnerable_validators.dedup();
+
+        <StakingInvulnerables<T>>::put(post_invulnerable_validators.clone());
+        weight = weight.saturating_add(RocksDbWeight::get().reads_writes(1, 1));
+
+        for valid_acc in &poa_validators {
+            if <StakingValidatorState<T>>::contains_key(&valid_acc) {
```

Comment:
I believe that's one read we need to account for

Reply:
True Good catch :+1: